### PR TITLE
Get pandas DataFrame for vertices and edges

### PR DESCRIPTION
**Pull request by iosonofabio**
Adding two methods to `Graph` to export vertices/edges and their names and attributes to pandas DataFrames. Basically vincent code from here:

https://igraph.discourse.group/t/convert-graph-object-to-dataframe/575

with a few changes for corner cases and tests.

Somehow my tests on 2.7 pass but on 3.9 complain they cannot import pandas. I have pandas in my python 3.9 installation, so... well let's see how CI goes.

### CODE REVIEW

**Review by ntamas — APPROVED** [review]

**Comment by ntamas** [comment]
Looks good to me! Is this ready for merge?

**Comment by vtraag** [comment]
I have a few comments, adding my review now.

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3232]: I don't think we ever use the `name` attribute for `edges` anywhere else, right? I think it might be strange to start using it here.

Instead, I would suggest that we add the `source` and `target` columns as index.

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3235]: It might make sense here to use the `name` of the `node_df`, because that is effectively also the `index` of the `vertex_df`.

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3236]: Likewise here of course.

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3193]: Perhaps we could make it optional to use the `name` attribute as the index or not? There are a few options here. We could allow the user to specify an attribute as index, or simply a boolean indicating or not whether the vertex `name` attribute should be used for creating the dataframe.

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3199]: Typo:
```suggestion
          The order of the vertices is the same as in the graph, so the i-th
```

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3218]: Typo:
```suggestion
          The order of the edges is the same as in the graph, so the i-th
```

File: src/igraph/__init__.py
Comment (vtraag) [inline src/igraph/__init__.py:3212]: Here, again, we might make it optional for the user to use either integer vertex indices, or the vertex labels (in the `source` and `target` columns).

**Review by vtraag — COMMENTED** [review]
Thanks for the implementation @iosonofabio, I have a couple of minor corrections, and some suggestions for improving it.

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3193]: The order of vertices is by id, not name, so one can currently index the vertex df by id or name as usual:

```python
df.iloc[myid]
df.loc[myname]
```

Most variations are one-liners, e.g.:

```python
df.index = np.arange(df.shape[0])  # From names to ids
df.set_index(mycol, inplace=True)  # From name/id to another column
```
so it's not hard for the user to achieve exactly what they want.

Instead, I would establish a decently consistent convention between nodes and edges, e.g.

1. using vertex ids as index, then in the source/target of the edge df we use vertex ids (this is my favourite)
2. using vertex names as index (current), then the source/target of the edge df use vertex names (less good because if the vertices have no names we get integers, otherwise strings, which tends to become a mess).

A bunch of ifs are not a problem, but I feel like telling the user about these one-liners in the docstring is worth more than baby-feeding them the exact variation.

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3199]: yep

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3218]: yep

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3212]: as above, I'd be more in favour of educating the user via docstrings than setting up a bunch of if/else for one-liners. Especially if that involves multiindices.

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3232]: Hmm... yes, the current version is just me being naive about the `name` attribute.

On the other hand, I am a little worried setting up a `pd.MultiIndex` with `(source, target)` columns as index. I use `pd.MultiIndex` a fair bit in my daily work, and it's kind of a pain. Say we want to get all incoming edges into vertex `i`:

```python
df.loc[df['target'] == i]  # without multi-index
df.query('target == i')  # without multi-index
df.loc[pd.IndexSlice[:, i], :]  # with multi-index
df.loc[df.index.get_level_values('target') == i]  # with multi-index
```
To me it looks like options 1 and 2 are easier to understand and actually mention the word `target`, while option 3 uses exotic objects (`IndexSlice`), exotic constructors via square brackets, and has a hidden requirement for specification of both row and column slicers (all those colons) which can otherwise fail silently (happened before). Option 4 is easier to understand but verbose.

I should have started replying to your review here, but alas I started below:

https://github.com/igraph/python-igraph/pull/355#discussion_r558615150

so what I wrote there is that ideally we wouldn't try to baby-feed one liners to our users. Having the multiindex is such a case:

```python
df.set_index(['source', 'target'], inplace=True)  # From columns to multi-index
df.reset_index(inplace=True)  # From multi-index to column, less well known I think
```
and if they use `drop=False` they can still slice using any of the 4 options. In such cases, I'd be in favour of documenting these transformations in the docstring rather than doing them for the user.

In any case this is just a slight preference, if you are strongly opinionated I'm happy to set up the multi-index too and document the one-liner required to snap out of it.

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3235]: This is a good point, that related to the https://github.com/igraph/python-igraph/pull/355#discussion_r558615150 below (why on Earth did I start replying there!)

Whichever we choose I agree the `source` and `target` cols/index levels should have the same thing as used to index the vertex dataframe, so that's not consistent at the moment.

I'd be tempted to advocate using vertex ids everywhere (i.e. keep this one and revert the vertex df to an index with just ids) because we know for a fact they are unique and well formatted. If we at some point start allowing people to use any Python hashables as vertex names, they will need to be here too which I'm a little afraid might cause a mess with dtypes.

But as for this whole PR, I'm only slightly opinionated, if you feel like you are using vertex names a lot more often than ids I'm happy to switch over everything to vertex names.

File: src/igraph/__init__.py
Comment (iosonofabio) [reply src/igraph/__init__.py:3236]: yep, see comment above

**Comment by iosonofabio** [comment]
Hi Vincent,

Now I've tried to do the opposite of what you asked for, i.e. simplified the functions to the bare minimum but added docstrings explaining the users how they can tweak the results. I thought it's easier seen than speculated about.

If you still think offering those output variations as keyword arguments is useful, I'm happy to add them and push another commit.

File: src/igraph/__init__.py
Comment (ntamas) [reply src/igraph/__init__.py:3216]: I agree with @iosonofabio here if that matters

File: src/igraph/__init__.py
Comment (ntamas) [reply src/igraph/__init__.py:3253]: > If we at some point start allowing people to use any Python hashables as vertex names,

oh please don't even think about that :) One has to draw a line somewhere.

Comment (iosonofabio) [reply src/igraph/__init__.py:3253]: We could do: anything that can be converted into a hashable without a loss of generality is accepted 😉

**Comment by iosonofabio** [comment]
I rebased to have hopefully a working CI. It should still fail because the tests are not adapted to the 'ID-only' policy proposed.

**Comment by ntamas** [comment]
I think that `pandas` is not installed in the CI environment, hence the failures.

**Comment by iosonofabio** [comment]
Would you like me to install it on CI or how are we going about testing this and other optional dependencies?

**Comment by ntamas** [comment]
First of al, you should use this to skip the test if Pandas is not installed:

```
try:
    import pandas as pd
except ImportError:
    pd = None

[... then later on in the test case ...]

    @skipIf(pd is None, "test case depends on Pandas")
    def testSomeStuffWithPandas(self):
        ...
```

Then you can add Pandas to `appveyor.yml` in the `install:` section after the `pip install cibuildwheel` step to make sure that the Pandas testcase gets a chance to execute in the CI.

**Comment by iosonofabio** [comment]
Alright, I will:
- double check all is working
- rebase on top of the `2to3` PR once that is merged
- lint/fix with black
- tag "ready for review"

**Comment by iosonofabio** [comment]
Alright, rebased. Tests are passed locally, let's wait for CI.

**Comment by iosonofabio** [comment]
I think this is good to go now

File: src/igraph/__init__.py
Comment (vtraag) [reply src/igraph/__init__.py]: Ha, I actually use this somewhere: I have a graph that determines how other graphs are coupled, see https://leidenalg.readthedocs.io/en/stable/multiplex.html#slices-to-layers.

But using such graphs as indices might become a bit more complicated for a pandas dataframe. In principle, pandas dataframes can simply store any Python object, as far as I know. I am not sure if there are any limitations on this for indices though?

I think it would make more sense to start working with names as much as possible (or at least try to support that if people want to), so for me, having this as an option would be good. Of course, it should somehow be consistent, but for that reason I think it should be left to a user to specify whether he wants the index to be with names, or with vertex integer identifiers. The default option should then be consistent with each other (and probably default to integers, or perhaps default to names only if names are available?)

File: src/igraph/__init__.py
Comment (vtraag) [reply src/igraph/__init__.py:3438]: Frankly, getting the source and target names are five lines of code, which I don't consider entirely trivial for end users. I personally never had too much trouble with `MultiIndex`, but maybe I just got lucky. But good if this is in the documentation.

**Review by vtraag — APPROVED** [review]

**Comment by ntamas** [comment]
Okay, Vincent approved so here we go now! Thanks a lot!

**Pull request merged**
